core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -0,0 +1,398 @@
+      Errors.INVALID_REQUEST
+    else if (!txnManager.isCoordinatorFor(transactionalId))
+      Errors.NOT_COORDINATOR
+    else if (txnManager.isCoordinatorLoadingInProgress(transactionalId))
+      Errors.COORDINATOR_LOAD_IN_PROGRESS
+    else
+      Errors.NONE
+
+
+  def handleAddPartitionsToTransaction(transactionalId: String,
+                                       pid: Long,
+                                       epoch: Short,
+                                       partitions: collection.Set[TopicPartition],
+                                       responseCallback: TxnMetadataUpdateCallback): Unit = {
+    val errors = validateTransactionalId(transactionalId)
+    if (errors != Errors.NONE)
+      responseCallback(errors)
+    else {
+      // try to update the transaction metadata and append the updated metadata to txn log;
+      // if there is no such metadata treat it as invalid pid mapping error.
+      val (error, newMetadata) = txnManager.getTransactionState(transactionalId) match {
+        case None =>
+          (Errors.INVALID_PID_MAPPING, null)
+
+        case Some(metadata) =>
+          // generate the new transaction metadata with added partitions
+          metadata synchronized {
+            if (metadata.pid != pid) {
+              (Errors.INVALID_PID_MAPPING, null)
+            } else if (metadata.producerEpoch != epoch) {
+              (Errors.INVALID_PRODUCER_EPOCH, null)
+            } else if (metadata.pendingState.isDefined) {
+              // return a retriable exception to let the client backoff and retry
+              (Errors.COORDINATOR_LOAD_IN_PROGRESS, null)
+            } else if (metadata.state != Empty && metadata.state != Ongoing) {
+              (Errors.INVALID_TXN_STATE, null)
+            } else if (partitions.subsetOf(metadata.topicPartitions)) {
+              // this is an optimization: if the partitions are already in the metadata reply OK immediately
+              (Errors.NONE, null)
+            } else {
+              val now = time.milliseconds()
+              val newMetadata = new TransactionMetadata(pid,
+                epoch,
+                metadata.txnTimeoutMs,
+                Ongoing,
+                metadata.topicPartitions ++ partitions,
+                now,
+                now)
+              metadata.prepareTransitionTo(Ongoing)
+              (Errors.NONE, newMetadata)
+            }
+          }
+      }
+
+      if (newMetadata != null) {
+        txnManager.appendTransactionToLog(transactionalId, newMetadata, responseCallback)
+      } else {
+        responseCallback(error)
+      }
+    }
+  }
+
+  def handleTxnImmigration(transactionStateTopicPartitionId: Int, coordinatorEpoch: Int) {
+    txnManager.loadTransactionsForPartition(transactionStateTopicPartitionId, coordinatorEpoch)
+  }
+
+  def handleTxnEmigration(transactionStateTopicPartitionId: Int) {
+    txnManager.removeTransactionsForPartition(transactionStateTopicPartitionId)
```

Comment:
Yes, I was referring to delayed events in the producer purgatory after the replicaManager.appendRecords(...) call. We can trigger a check of those delayed events and let them error out (since the check will find out that the broker is no longer the coordinator).